### ISSUE TITLE
[bug]  assert len(modules) == len(inputs) when use torch.distributed to compute last batch

### ISSUE THREAD

**Issue by piaozhx** [pytorch/pytorch#5587]
PyTorch GitHub Issues Guidelines
--------------------------------
- OS: CentOS 7.2.1511
- PyTorch version: 0.3.1 post2
- How you installed PyTorch (conda, pip, source): pip
- Python version:3.6.3 
- CUDA/cuDNN version: cuda8
- GPU models and configuration: Tesla P40

Recently, I use torch.distributed to run my model on multi node(each node have 4 P40), sometime(e.g. use 3 processes on corresponding nodes, each process use 4 cards) It works well, but most time(e.g. use 4 nodes, 5 nodes ,...) I get a strange error:

```
Traceback (most recent call last):
 File "triplet_dist.py", line 358, in <module>
   model.main()
 File "triplet_dist.py", line 332, in main
   self.validate(epoch)
 File "triplet_dist.py", line 243, in validate
   outputs, dist_a, dist_b, embed_x, embed_y, embed_z = self.model(inputs, input0, fgs, bgs)
 File "/public/home/gaoshenghua/zhengjia/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 357, in __call__
   result = self.forward(*input, **kwargs)
 File "/public/home/gaoshenghua/zhengjia/anaconda3/lib/python3.6/site-packages/torch/nn/parallel/distributed.py", line 167, in forward
   outputs = self.parallel_apply(self._module_copies, inputs, kwargs)
 File "/public/home/gaoshenghua/zhengjia/anaconda3/lib/python3.6/site-packages/torch/nn/parallel/distributed.py", line 174, in parallel_apply
   return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
 File "/public/home/gaoshenghua/zhengjia/anaconda3/lib/python3.6/site-packages/torch/nn/parallel/parallel_apply.py", line 24, in parallel_apply
   assert len(modules) == len(inputs)
AssertionError
terminate called after throwing an instance of 'gloo::EnforceNotMet'
 what():  [enforce fail at /opt/conda/conda-bld/pytorch_1518243271935/work/torch/lib/gloo/gloo/cuda_private.h:40] error == cudaSuccess. 29 vs 0. Error at: /opt/conda/conda-bld/pytorch_1518243271935/work/torch/lib/gloo/gloo/cuda_private.h:40: driver shutting down
``` 

I found that error always happened when the model compute last batch of a epoch (for example, total dataset size is 100, batch size is 30, then I have 4 batch but last batch size is 10) ,If I just dropped last batch, everything will be OK. but I think "just drop it" is very dirty and is impossible for test dataset. Any good idea for that case?

**Comment by apaszke** [comment]
As a quick fix you can pass `drop_last=True` when creating the data loader, but it's a bug that we'll need to fix

**Comment by lliimsft** [comment]
Hi Adam, I set the drop_last to True, but still get the same error with GPU+Gloo

**Comment by richielo** [comment]
May I ask if this fix has been included in the 0.4.1 version of PyTorch? I am still facing the same issue using DataParallel. I am using torchtext's Iterator instead of Dataloader though. Thank you very much
